multicluster/controllers/multicluster/label_identity_export_controller.go

```diff
@@ -0,0 +1,321 @@
+		}
+	}
+	id, err := r.allocator.allocate()
+	if err != nil {
+		klog.ErrorS(err, "Failed to allocate ID for new labels", "label", labelHash)
+		return false
+	}
+	labelIdentityResImport = getLabelIdentityResImport(labelHash, label, namespace, id)
+	if err := r.Client.Create(ctx, labelIdentityResImport, &client.CreateOptions{}); err != nil {
+		klog.ErrorS(err, "Failed to create LabelIdentity kind of ResourceImport for new label", "label", label)
+		r.allocator.release(id)
+		return false
+	}
+	r.labelsToID[labelHash] = id
+	if clusters, ok := r.labelsToClusters[labelHash]; ok {
+		clusters.Insert(clusterID)
+	} else {
+		r.labelsToClusters[labelHash] = sets.NewString(clusterID)
+	}
+	if labels, ok := r.clusterToLabels[clusterID]; ok {
+		labels.Insert(labelHash)
+	} else {
+		r.clusterToLabels[clusterID] = sets.NewString(labelHash)
+	}
+	return true
+}
+
+func getLabelIdentityResImport(labelHash, label, ns string, id uint32) *mcsv1alpha1.ResourceImport {
+	return &mcsv1alpha1.ResourceImport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      labelHash,
+			Namespace: ns,
+		},
+		Spec: mcsv1alpha1.ResourceImportSpec{
+			LabelIdentity: &mcsv1alpha1.LabelIdentitySpec{
+				Label: label,
+				ID:    id,
+			},
+		},
+	}
+}
+
+// parseLabelIdentityExportNamespacedName gets the clusterID and label identity
+// hash from the API request.
+func parseLabelIdentityExportNamespacedName(namespacedName types.NamespacedName) (string, string) {
+	lastIdx := strings.LastIndex(namespacedName.Name, "-")
+	clusterID := namespacedName.Name[:lastIdx]
+	labelHash := namespacedName.Name[lastIdx+1:]
+	return clusterID, labelHash
+}
+
+// idAllocator allocates a unqiue uint32 ID for each label identity.
+type idAllocator struct {
+	sync.Mutex
+	maxID                uint32
+	nextID               uint32
+	preAllocatedIDs      intsets.Sparse
+	availableForReuseIDs intsets.Sparse
+}
+
+// allocate will first try to allocate an ID within the pool of IDs that has been
+// released (due to label identity deletions). If there's no such IDs, it will
+// then allocate the first ID that has not been pre-allocated, or return an error
+// if all IDs have been exhausted.
+func (a *idAllocator) allocate() (uint32, error) {
+	a.Lock()
+	defer a.Unlock()
+
+	available := -1
+	if ok := a.availableForReuseIDs.TakeMin(&available); ok {
+		return uint32(available), nil
+	}
+	for a.preAllocatedIDs.Has(int(a.nextID)) {
+		a.nextID += 1
+	}
+	if a.nextID <= a.maxID {
+		allocated := a.nextID
+		a.nextID += 1
+		return allocated, nil
+	}
+	return 0, fmt.Errorf("no ID available")
+}
+
+// setAllocated reserves IDs allocated during the previous round of label identity
+// ResourceExport reconcilaion.
+func (a *idAllocator) setAllocated(id uint32) error {
+	a.Lock()
+	defer a.Unlock()
+
+	if a.availableForReuseIDs.Has(int(id)) {
+		a.availableForReuseIDs.Remove(int(id))
+		return nil
+	} else if id >= a.nextID {
+		a.preAllocatedIDs.Insert(int(id))
```

Comment:
I did not get why we need preAllocatedIDs. Could you explain?

Reply:
In case of leader multi-cluster controller restart, the cache between label identities and their corresponding allocated IDs will be lost. However, any ResourceImports (which contain these mappings) and ACNPs (which are enforced based on the IDs) will still exist in the ClusterSet. This is different from networkpolicy controller, where for each agent controller epoch we wipe out all OVS flows and install new ones. We do not want to erase all ResourceImports and/or ACNP during multi-cluster controller restart as it can cause dataplane churn. Hence, upon restart the mcs controller should "learn" about IDs allocated for label identities in the previous epoch and set these as preAllocated, unless the label identity is deleted thereafter and then will be put into the "available for reuse" pool.

Reply:
I got that we need to restore the state after restart, but I did not understand why we cannot rebuild the state the same as normal allocation. Maybe it is too complex to rebuild availableForReuseIDs and nextID from the discovered IDs?

Also feel "preAllocatedIDs" is a little misleading. Do you think something like "restoredIDs" or "restoredAllocatedIDs" works better? And "availableForReuseIDs" to "releasedIDs"?

Reply:
Thanks for the naming suggestion, will update. I won't say it's "complex" to rebuild the allocator internals from the discovered IDs, we just need a synchronized solution. i.e. upon controller restart, it pulls all Label Identity ResourceImports and restore the IDs in batch. Right now this restoration process happens asynchronously. @tnqn for suggestions, and also whether we shall share the allocator with other modules.